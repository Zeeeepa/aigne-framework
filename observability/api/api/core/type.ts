import type { InferInsertModel } from "drizzle-orm";
import { z } from "zod";
import type { Trace } from "../server/models/trace.js";

export const AIGNEObserverOptionsSchema = z
  .object({ storage: z.string().optional() })
  .optional()
  .default({});

export type AIGNEObserverOptions = z.infer<typeof AIGNEObserverOptionsSchema>;

export type TraceFormatSpans = Omit<InferInsertModel<typeof Trace>, "id" | "rootId"> & {
  id: string;
  rootId: string;
};
