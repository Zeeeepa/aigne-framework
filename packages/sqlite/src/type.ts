import { customType } from "drizzle-orm/sqlite-core";

export const json = customType<{
  data: unknown;
  driverData: string;
}>({
  dataType: () => "json",
  fromDriver: (value) => JSON.parse(value),
  toDriver: (value) => JSON.stringify(value),
});

export const datetime = customType<{
  data: Date;
  driverData: string;
}>({
  dataType: () => "datetime",
  fromDriver: (value) => new Date(value),
  toDriver: (value) => value.toISOString(),
});
