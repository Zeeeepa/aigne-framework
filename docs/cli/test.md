# Testing

Use AIGNE CLI to run agent tests.

## `aigne test` Command

Run tests in the specified agent directory.

```bash
aigne test [options]
```

### Options

* `--path <path>`: Path to the agent directory (defaults to current directory `.`)
* `--help`: Show command help

### Examples

#### Run tests in current directory

```bash
aigne test
```

#### Run tests in specific directory

```bash
aigne test --path ./my-agents
```

### Test File Structure

The AIGNE testing system automatically finds and runs test files in your project. Make sure your test files follow the correct naming conventions and structure.

### Best Practices

1. **Organize test files**: Place test files in appropriate directory structures
2. **Write clear test cases**: Ensure each test case has a clear purpose and expected outcome
3. **Use descriptive test names**: Make test names clearly express what the test covers
4. **Run tests regularly**: Run tests regularly during development to ensure code quality
