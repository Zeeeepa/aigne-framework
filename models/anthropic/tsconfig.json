{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "paths": {
      "@aigne/anthropic": ["./src"],
      "@aigne/anthropic/*": ["./src/*"]
    }
  },
  "exclude": ["node_modules", "lib"],
  "include": ["src", "test"]
}
