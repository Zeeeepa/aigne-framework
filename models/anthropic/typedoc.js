import base from "../../typedoc.base.js";

/**
 * @type {import('typedoc').TypeDocOptions}
 */
export default {
  ...base,
  entryPoints: ["src/index.ts"],
};
