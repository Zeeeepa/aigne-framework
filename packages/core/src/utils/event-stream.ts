import { createParser } from "eventsource-parser";
import { produce } from "immer";
import {
  type AgentResponseChunk,
  type AgentResponseProgress,
  type AgentResponseStream,
  isAgentResponseDelta,
  isAgentResponseProgress,
  type Message,
} from "../agents/agent.js";
import type { Context, ContextEventMap } from "../aigne/context.js";
import { tryOrThrow } from "./type-utils.js";
import type { Listener } from "./typed-event-emitter.js";

export class EventStreamParser<T> extends TransformStream<string, T | Error> {
  constructor() {
    let parser: ReturnType<typeof createParser> | undefined;

    super({
      start(controller) {
        parser = createParser({
          onEvent: (event) => {
            const json = tryOrThrow(
              () => JSON.parse(event.data) as T | { message: string },
              (e) => {
                controller.enqueue(
                  new Error(`Parse response chunk json error: ${e.message} ${event.data}`),
                );
              },
            );
            if (json) {
              switch (event.event) {
                case "error":
                  controller.enqueue(new Error((json as { message: string }).message));
                  break;
                default: {
                  if (!event.event) controller.enqueue(json as T);
                  else console.warn(`Unknown event type: ${event.event}`, event.data);
                }
              }
            }
          },
        });
      },
      transform(chunk) {
        parser?.feed(chunk);
      },
    });
  }
}

export class AgentResponseStreamParser<O extends Message> extends TransformStream<
  AgentResponseChunk<O> | Error,
  AgentResponseChunk<O>
> {
  private json: O = {} as O;

  constructor() {
    super({
      transform: (chunk, controller) => {
        if (chunk instanceof Error) {
          controller.error(chunk);
          controller.terminate();
          return;
        }

        if (isAgentResponseDelta(chunk)) {
          this.json = produce(this.json, (draft) => {
            if (chunk.delta.json) Object.assign(draft, chunk.delta.json);

            if (chunk.delta.text) {
              for (const [key, text] of Object.entries(chunk.delta.text)) {
                const original = draft[key] as string | undefined;
                const t = (original || "") + (text || "");
                if (t) Object.assign(draft, { [key]: t });
              }
            }
          });

          controller.enqueue({
            ...chunk,
            delta: {
              ...chunk.delta,
              json: this.json,
            },
          });
        } else if (isAgentResponseProgress(chunk)) {
          if (chunk.progress.event === "agentFailed") {
            const { name, message } = chunk.progress.error;
            chunk.progress.error = new Error(message);
            chunk.progress.error.name = name;
          }
          controller.enqueue(chunk);
        }
      },
    });
  }
}

export class AgentResponseStreamSSE<O extends Message> extends ReadableStream<string> {
  constructor(stream: AgentResponseStream<O>) {
    let reader: ReadableStreamDefaultReader<AgentResponseChunk<O>> | undefined;

    super({
      async pull(controller) {
        reader ??= stream.getReader();
        try {
          const { value, done } = await reader.read();
          if (done) {
            controller.close();
            return;
          }

          if (isAgentResponseProgress(value)) {
            if (value.progress.event === "agentFailed") {
              value.progress.error = {
                name: value.progress.error.name,
                message: value.progress.error.message,
              };
            }
          }

          controller.enqueue(`data: ${JSON.stringify(value)}\n\n`);
        } catch (error) {
          controller.enqueue(
            `event: error\ndata: ${JSON.stringify({ message: error.message })}\n\n`,
          );
          controller.close();
        }
      },
    });
  }
}

export class AgentResponseProgressStream extends ReadableStream<AgentResponseProgress> {
  constructor(context: Context) {
    super({
      async start(controller) {
        const writeEvent = (
          eventName: keyof ContextEventMap,
          event: ContextEventMap[typeof eventName][0],
        ) => {
          const progress = {
            ...event,
            event: eventName,
            agent: { name: event.agent.name },
          } as AgentResponseProgress["progress"];

          controller.enqueue({ progress });
        };

        const close = () => {
          context.off("agentStarted", onAgentStarted);
          context.off("agentSucceed", onAgentSucceed);
          context.off("agentFailed", onAgentFailed);
          controller.close();
        };

        const onAgentStarted: Listener<"agentStarted", ContextEventMap> = async (event) => {
          const taskTitle = await event.agent.renderTaskTitle(event.input);

          writeEvent("agentStarted", { ...event, taskTitle });
        };
        const onAgentSucceed: Listener<"agentSucceed", ContextEventMap> = (event) => {
          writeEvent("agentSucceed", event);
          if (event.contextId === context.id) {
            close();
          }
        };
        const onAgentFailed: Listener<"agentFailed", ContextEventMap> = (event) => {
          writeEvent("agentFailed", event);
          if (event.contextId === context.id) {
            close();
          }
        };

        context.on("agentStarted", onAgentStarted);
        context.on("agentSucceed", onAgentSucceed);
        context.on("agentFailed", onAgentFailed);
      },
    });
  }
}
