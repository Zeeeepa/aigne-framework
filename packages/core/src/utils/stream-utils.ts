import equal from "fast-deep-equal";
import {
  type AgentProcessAsyncGenerator,
  type AgentResponseChunk,
  type AgentResponseStream,
  isAgentResponseDelta,
  isEmptyChunk,
  type Message,
} from "../agents/agent.js";
import { isNonNullable, isRecord, omitBy, type PromiseOrValue } from "./type-utils.js";
import "./stream-polyfill.js";
import type { ReadableStreamDefaultReadResult } from "bun";

export function objectToAgentResponseStream<T extends Message>(json: T): AgentResponseStream<T> {
  if (!isRecord(json)) {
    throw new Error(
      `expect to return a record type such as {result: ...}, but got (${typeof json}): ${json}`,
    );
  }

  return new ReadableStream({
    pull(controller) {
      controller.enqueue({ delta: { json } });
      controller.close();
    },
  });
}

export function mergeAgentResponseChunk<T extends Message>(
  output: T,
  chunk: AgentResponseChunk<T>,
) {
  if (isAgentResponseDelta(chunk)) {
    if (chunk.delta.text) {
      for (const [key, text] of Object.entries(chunk.delta.text)) {
        const original = output[key] as string | undefined;
        const t = (original || "") + (text || "");
        if (t) Object.assign(output, { [key]: t });
      }
    }

    if (chunk.delta.json) {
      Object.assign(
        output,
        omitBy(chunk.delta.json, (v) => v === undefined),
      );
    }
  }

  return output;
}

export async function agentResponseStreamToObject<T extends Message>(
  stream: AgentResponseStream<T> | AgentProcessAsyncGenerator<T>,
): Promise<T> {
  const json: T = {} as T;

  if (stream instanceof ReadableStream) {
    for await (const value of stream) {
      mergeAgentResponseChunk(json, value);
    }
  } else {
    for (;;) {
      const chunk = await stream.next();
      if (chunk.value) {
        if (chunk.done) {
          Object.assign(json, chunk.value);
        } else {
          mergeAgentResponseChunk(json, chunk.value);
        }
      }
      if (chunk.done) break;
    }
  }

  return json;
}

export function asyncGeneratorToReadableStream<T extends Message>(
  generator: AgentProcessAsyncGenerator<T>,
): AgentResponseStream<T> {
  return new ReadableStream({
    async pull(controller) {
      try {
        const chunk = await generator.next();
        if (chunk.value) {
          if (chunk.done) {
            controller.enqueue({ delta: { json: chunk.value } });
          } else {
            controller.enqueue(chunk.value);
          }
        }

        if (chunk.done) {
          controller.close();
        }
      } catch (error) {
        controller.error(error);
      }
    },
  });
}

export function onAgentResponseStreamEnd<T extends Message>(
  stream: AgentResponseStream<T>,
  options?: {
    onChunk?: (
      chunk: AgentResponseChunk<T>,
    ) => PromiseOrValue<AgentResponseChunk<T> | undefined | void>;
    onResult?: (result: T) => PromiseOrValue<Partial<T> | undefined | void>;
    onError?: (error: Error) => PromiseOrValue<Error>;
  },
): AgentResponseStream<T> {
  const json: T = {} as T;
  const reader = stream.getReader();

  return new ReadableStream({
    async pull(controller) {
      try {
        while (true) {
          const { value, done } = await reader.read();

          if (done) {
            const result = (await options?.onResult?.(json)) ?? json;

            if (result && !equal(result, json)) {
              let chunk: AgentResponseChunk<T> = { delta: { json: result } };
              chunk = (await options?.onChunk?.(chunk)) ?? chunk;
              controller.enqueue(chunk);
            }

            controller.close();

            return;
          }

          mergeAgentResponseChunk(json, value);

          const chunk = (await options?.onChunk?.(value)) ?? value;
          if (!isEmptyChunk(chunk)) {
            controller.enqueue(chunk);
            break;
          }
        }
      } catch (error) {
        controller.error((await options?.onError?.(error)) ?? error);
      }
    },
  });
}

export function isAsyncGenerator<T extends AsyncGenerator>(
  value: AsyncGenerator | unknown,
): value is T {
  return typeof value === "object" && value !== null && Symbol.asyncIterator in value;
}

export async function* arrayToAgentProcessAsyncGenerator<T extends Message>(
  chunks: (AgentResponseChunk<T> | Error)[],
  result?: Partial<T>,
): AgentProcessAsyncGenerator<T> {
  for (const chunk of chunks) {
    if (chunk instanceof Error) throw chunk;

    yield chunk;
  }
  if (result !== undefined) return result;
}

export function arrayToReadableStream<T>(chunks: (T | Error)[]): ReadableStream<T> {
  const list = [...chunks];

  return new ReadableStream({
    pull(controller) {
      const item = list.shift();
      if (!item) {
        controller.close();
        return;
      }

      if (item instanceof Error) {
        controller.error(item);
        return;
      }

      controller.enqueue(item);
    },
  });
}

export async function readableStreamToArray<T>(
  stream: ReadableStream<T>,
  options: { catchError: true },
): Promise<(T | Error)[]>;
export async function readableStreamToArray<T>(
  stream: ReadableStream<T>,
  options?: { catchError?: boolean },
): Promise<T[]>;
export async function readableStreamToArray<T>(
  stream: ReadableStream<T>,
  options?: { catchError?: boolean },
): Promise<(T | Error)[]> {
  const result: T[] = [];
  try {
    for await (const value of stream) {
      result.push(value);
    }
  } catch (error) {
    if (!options?.catchError) throw error;

    result.push(error);
  }
  return result;
}

export function stringToAgentResponseStream(
  str: string,
  key: "text" | string = "text",
): AgentResponseStream<Message> {
  const segmenter = new Intl.Segmenter(undefined, { granularity: "word" });
  const segments = segmenter.segment(str);

  return arrayToReadableStream(
    Array.from(segments).map((segment) => ({ delta: { text: { [key]: segment.segment } } })),
  );
}

export function toReadableStream(stream: NodeJS.ReadStream) {
  return new ReadableStream<Uint8Array>({
    start(controller) {
      const onData = (chunk: Buffer) => {
        controller.enqueue(new Uint8Array(chunk));
      };

      const onEnd = () => {
        cleanup();
        controller.close();
      };

      const onError = (err: Error) => {
        cleanup();
        controller.error(err);
      };

      function cleanup() {
        stream.off("data", onData);
        stream.off("end", onEnd);
        stream.off("error", onError);
      }

      stream.on("data", onData);
      stream.on("end", onEnd);
      stream.on("error", onError);
    },
  });
}

export async function readAllString(stream: NodeJS.ReadStream | ReadableStream): Promise<string> {
  return (
    await readableStreamToArray(
      (stream instanceof ReadableStream ? stream : toReadableStream(stream)).pipeThrough(
        new TextDecoderStream(),
      ),
    )
  ).join("");
}

export function mergeReadableStreams<T1, T2>(
  s1: ReadableStream<T1>,
  s2: ReadableStream<T2>,
): ReadableStream<T1 | T2>;
export function mergeReadableStreams(
  ...streams: (ReadableStream<any> | undefined)[]
): ReadableStream<any>;
export function mergeReadableStreams(
  ...streams: (ReadableStream<any> | undefined)[]
): ReadableStream<any> {
  type Reader = {
    reader: ReadableStreamDefaultReader;
    reading?: Promise<{
      result: ReadableStreamDefaultReadResult<any>;
      item: any;
    }>;
  };

  let readers: Reader[] | undefined;

  return new ReadableStream({
    async pull(controller) {
      try {
        readers ??= streams.filter(isNonNullable).map((s) => ({ reader: s.getReader(), data: [] }));

        while (readers.length) {
          const chunk: Awaited<NonNullable<Reader["reading"]>> = await Promise.race(
            readers.map((i) => {
              i.reading ??= i.reader.read().then((result) => ({ result, item: i }));
              return i.reading;
            }),
          );

          if (chunk.result.value) {
            controller.enqueue(chunk.result.value);
            chunk.item.reading = undefined;
            return;
          }

          if (chunk.result.done) {
            readers = readers.filter((i) => i !== chunk.item);
          }
        }

        controller.close();
      } catch (error) {
        controller.error(error);
        if (readers) for (const item of readers) item.reader.releaseLock();
      }
    },
  });
}
