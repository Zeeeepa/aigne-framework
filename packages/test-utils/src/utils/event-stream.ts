import { readFile } from "node:fs/promises";

export function createMockEventStream<T>(data: { path: string } | { raw: string }): T {
  return new ReadableStream({
    async start(controller) {
      const file = "path" in data ? await readFile(data.path) : data.raw;
      for (const line of file.toString().split("\n")) {
        if (line) controller.enqueue(JSON.parse(line.replace("data:", "")));
      }
      controller.close();
    },
  }) as unknown as T;
}
