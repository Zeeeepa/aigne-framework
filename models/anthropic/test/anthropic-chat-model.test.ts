import { expect, spyOn, test } from "bun:test";
import { join } from "node:path";
import { AnthropicChatModel } from "@aigne/anthropic";
import { isAgentResponseDelta, textDelta } from "@aigne/core";
import { readableStreamToArray } from "@aigne/core/utils/stream-utils.js";
import { createMockEventStream } from "@aigne/test-utils/utils/event-stream.js";
import {
  COMMON_RESPONSE_FORMAT,
  COMMON_TOOLS,
  createWeatherToolCallMessages,
  createWeatherToolMessages,
} from "@aigne/test-utils/utils/openai-like-utils.js";
import type Anthropic from "@anthropic-ai/sdk";

test("Anthropic chat model basic usage", async () => {
  // #region example-anthropic-chat-model
  const model = new AnthropicChatModel({
    // Provide API key directly or use environment variable ANTHROPIC_API_KEY or CLAUDE_API_KEY
    apiKey: "your-api-key", // Optional if set in env variables
    // Specify Claude model version (defaults to 'claude-3-7-sonnet-latest')
    model: "claude-3-haiku-20240307",
    // Configure model behavior
    modelOptions: {
      temperature: 0.7,
    },
  });

  spyOn(model, "process").mockReturnValueOnce({
    text: "I'm Claude, an AI assistant created by Anthropic. How can I help you today?",
    model: "claude-3-haiku-20240307",
    usage: {
      inputTokens: 8,
      outputTokens: 15,
    },
  });

  const result = await model.invoke({
    messages: [{ role: "user", content: "Tell me about yourself" }],
  });

  console.log(result);
  /* Output:
  {
    text: "I'm Claude, an AI assistant created by Anthropic. How can I help you today?",
    model: "claude-3-haiku-20240307",
    usage: {
      inputTokens: 8,
      outputTokens: 15
    }
  }
  */

  expect(result).toEqual({
    text: "I'm Claude, an AI assistant created by Anthropic. How can I help you today?",
    model: "claude-3-haiku-20240307",
    usage: {
      inputTokens: 8,
      outputTokens: 15,
    },
  });
  // #endregion example-anthropic-chat-model
});

test("Anthropic chat model with streaming using async generator", async () => {
  // #region example-anthropic-chat-model-streaming-async-generator
  const model = new AnthropicChatModel({
    apiKey: "your-api-key",
    model: "claude-3-haiku-20240307",
  });

  spyOn(model, "process").mockImplementationOnce(async function* () {
    yield textDelta({ text: "I'm Claude" });
    yield textDelta({ text: ", an AI assistant" });
    yield textDelta({ text: " created by Anthropic." });
    yield textDelta({ text: " How can I help you today?" });

    return {
      model: "claude-3-haiku-20240307",
      usage: { inputTokens: 8, outputTokens: 15 },
    };
  });

  const stream = await model.invoke(
    {
      messages: [{ role: "user", content: "Tell me about yourself" }],
    },
    { streaming: true },
  );

  let fullText = "";
  const json = {};

  for await (const chunk of stream) {
    if (isAgentResponseDelta(chunk)) {
      const text = chunk.delta.text?.text;
      if (text) fullText += text;
      if (chunk.delta.json) Object.assign(json, chunk.delta.json);
    }
  }

  console.log(fullText); // Output: "I'm Claude, an AI assistant created by Anthropic. How can I help you today?"
  console.log(json); // { model: "claude-3-haiku-20240307", usage: { inputTokens: 8, outputTokens: 15 } }

  expect(fullText).toBe(
    "I'm Claude, an AI assistant created by Anthropic. How can I help you today?",
  );
  expect(json).toEqual({
    model: "claude-3-haiku-20240307",
    usage: { inputTokens: 8, outputTokens: 15 },
  });
  // #endregion example-anthropic-chat-model-streaming-async-generator
});

test("AnthropicChatModel.invoke", async () => {
  const model = new AnthropicChatModel({
    apiKey: "YOUR_API_KEY",
  });

  spyOn(model.client.messages, "stream")
    .mockReturnValueOnce(
      createMockEventStream({
        path: join(import.meta.dirname, "anthropic-streaming-response-1.txt"),
      }),
    )
    .mockReturnValueOnce(
      createMockEventStream({
        path: join(import.meta.dirname, "anthropic-streaming-response-2.txt"),
      }),
    );

  spyOn(model.client.messages, "create").mockResolvedValue(
    (await import(
      "./anthropic-structured-response-3.json"
    )) as unknown as Anthropic.Messages.Message,
  );

  const result1 = await model.invoke({
    messages: createWeatherToolMessages(),
    tools: COMMON_TOOLS,
    responseFormat: COMMON_RESPONSE_FORMAT,
  });

  expect(result1).toMatchSnapshot();

  const result2 = await model.invoke({
    messages: createWeatherToolCallMessages(),
    tools: COMMON_TOOLS,
    responseFormat: COMMON_RESPONSE_FORMAT,
  });

  expect(result2).toMatchSnapshot();
});

test("AnthropicChatModel.invoke should pass system and messages to claude correctly", async () => {
  const model = new AnthropicChatModel({
    apiKey: "YOUR_API_KEY",
  });

  const stream = spyOn(model.client.messages, "stream").mockReturnValueOnce(
    createMockEventStream({
      path: join(import.meta.dirname, "anthropic-streaming-response-1.txt"),
    }),
  );

  await model.invoke({
    messages: [
      { role: "system", content: "You are a chatbot" },
      { role: "user", content: "hello" },
    ],
  });

  expect(stream.mock.calls).toEqual([
    [
      expect.objectContaining({
        system: "You are a chatbot",
        messages: [{ role: "user", content: "hello" }],
      }),
    ],
  ]);
});

test("AnthropicChatModel.invoke should use system message as user message if messages is empty", async () => {
  const model = new AnthropicChatModel({
    apiKey: "YOUR_API_KEY",
  });

  const stream = spyOn(model.client.messages, "stream").mockReturnValueOnce(
    createMockEventStream({
      path: join(import.meta.dirname, "anthropic-streaming-response-1.txt"),
    }),
  );

  await model.invoke({
    messages: [{ role: "system", content: "You are a chatbot" }],
  });

  // Should not include system property in the request, the system message should be treated as user message
  expect(stream.mock.calls).toEqual([
    [
      expect.not.objectContaining({
        system: expect.anything(),
      }),
    ],
  ]);
  expect(stream.mock.calls).toEqual([
    [
      expect.objectContaining({
        messages: [{ role: "user", content: "You are a chatbot" }],
      }),
    ],
  ]);
});

test("AnthropicChatModel.invoke with streaming", async () => {
  const model = new AnthropicChatModel({
    apiKey: "YOUR_API_KEY",
  });

  spyOn(model.client.messages, "stream").mockReturnValueOnce(
    createMockEventStream({
      path: join(import.meta.dirname, "anthropic-streaming-response-text.txt"),
    }),
  );

  const stream = await model.invoke(
    {
      messages: [{ role: "user", content: "hello" }],
    },
    { streaming: true },
  );

  expect(readableStreamToArray(stream)).resolves.toMatchSnapshot();
});

test("AnthropicChatModel.invoke without streaming", async () => {
  const model = new AnthropicChatModel({
    apiKey: "YOUR_API_KEY",
  });

  spyOn(model.client.messages, "stream").mockReturnValue(
    createMockEventStream({
      path: join(import.meta.dirname, "anthropic-streaming-response-text.txt"),
    }),
  );

  const result = await model.invoke({
    messages: [{ role: "user", content: "hello" }],
  });

  expect(result).toMatchSnapshot();
});
