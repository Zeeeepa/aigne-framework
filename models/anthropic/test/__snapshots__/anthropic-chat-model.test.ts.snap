// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`AnthropicChatModel.invoke 1`] = `
{
  "model": "claude-3-7-sonnet-20250219",
  "toolCalls": [
    {
      "function": {
        "arguments": {
          "city": "New York",
        },
        "name": "get_weather",
      },
      "id": "toolu_01MnGRVuuum7T1AiVoRh89QC",
      "type": "function",
    },
  ],
  "usage": {
    "inputTokens": 416,
    "outputTokens": 54,
  },
}
`;

exports[`AnthropicChatModel.invoke 2`] = `
{
  "json": {
    "text": "The current temperature in New York is 20 degrees.",
  },
  "model": "claude-3-7-sonnet-20250219",
  "usage": {
    "inputTokens": 486,
    "outputTokens": 20,
  },
}
`;

exports[`AnthropicChatModel.invoke with streaming 1`] = `
[
  {
    "delta": {
      "json": {
        "model": "claude-3-7-sonnet-20250219",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "Hello! How",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " can I assist you today? Whether you have a question,",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " need information, or just want to chat, I'm here",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " to help. Feel free to let me know what",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "'s on your mind.",
      },
    },
  },
  {
    "delta": {
      "json": {
        "usage": {
          "inputTokens": 8,
          "outputTokens": 45,
        },
      },
    },
  },
]
`;

exports[`AnthropicChatModel.invoke without streaming 1`] = `
{
  "model": "claude-3-7-sonnet-20250219",
  "text": "Hello! How can I assist you today? Whether you have a question, need information, or just want to chat, I'm here to help. Feel free to let me know what's on your mind.",
  "usage": {
    "inputTokens": 8,
    "outputTokens": 45,
  },
}
`;
